rasa/telemetry.py

```diff
@@ -0,0 +1,553 @@
+        "cpu": multiprocessing.cpu_count(),
+        "docker": _is_docker(),
+    }
+
+
+@ensure_telemetry_enabled
+async def track(
+    event_name: Text,
+    properties: Optional[Dict[Text, Any]] = None,
+    context: Optional[Dict[Text, Any]] = None,
+) -> None:
+    """Tracks a telemetry event.
+
+    It is OK to use this function from outside telemetry.py, but note that it
+    is recommended to create a new track_xyz() function for complex telemetry
+    events, or events that are generated from many parts of the Rasa Open Source code.
+
+    Args:
+        event_name: Name of the event.
+        properties: Dictionary containing the event's properties.
+        context: Dictionary containing some context for this event.
+    """
+
+    telemetry_id = get_telemetry_id()
+
+    if not telemetry_id:
+        logger.debug("Will not report telemetry events as no ID was found.")
+        return
+
+    if not properties:
+        properties = {}
+
+    properties[TELEMETRY_ID] = telemetry_id
+
+    await _send_event(
+        telemetry_id, event_name, properties, with_default_context_fields(context)
+    )
+
+
+def get_telemetry_id() -> Optional[Text]:
+    """Return the unique telemetry identifier for this Rasa Open Source install.
+    The identifier can be any string, but it should be a UUID.
+
+    Returns:
+        The identifier, if it is configured correctly.
+    """
+
+    try:
+        telemetry_config = (
+            rasa_utils.read_global_config_value(CONFIG_FILE_TELEMETRY_KEY) or {}
+        )
+
+        return telemetry_config.get(CONFIG_TELEMETRY_ID)
+    except Exception as e:  # skipcq:PYL-W0703
+        logger.debug(f"Unable to retrieve telemetry ID: {e}")
+        return None
+
+
+def toggle_telemetry_reporting(is_enabled: bool) -> None:
+    """Write to the configuration if telemetry tracking should be enabled or disabled.
+
+    Args:
+        is_enabled: `True` if the telemetry reporting should be enabled,
+            `False` otherwise.
+    """
+
+    configuration = rasa_utils.read_global_config_value(CONFIG_FILE_TELEMETRY_KEY)
+
+    if configuration:
+        configuration[CONFIG_TELEMETRY_ENABLED] = is_enabled
+    else:
+        configuration = _default_telemetry_configuration(is_enabled)
+
+    rasa_utils.write_global_config_value(CONFIG_FILE_TELEMETRY_KEY, configuration)
+
+
+@ensure_telemetry_enabled
+@async_generator.asynccontextmanager
+async def track_model_training(
+    training_data: TrainingDataImporter, model_type: Text
+) -> None:
+    """Track a model training started.
+
+    Args:
+        training_data: Training data used for the training.
+        model_type: Specifies the type of training, should be either "rasa", "core"
+            or "nlu".
+    """
+
+    config = await training_data.get_config()
+    stories = await training_data.get_stories()
+    nlu_data = await training_data.get_nlu_data()
+    domain = await training_data.get_domain()
```

Comment:
it doesn't matter but should be `asyncio.gather` in case it does matter at one point

Reply:
nah, doesn't matter - this is easier to read 🙂

Reply:
But what if a community member implements an importer that for example retrieves training data from a server? I don't think readability is such a big thing if you unpack it immediately after calling `gather(...)`

```
config, stories, nlu_data, domain = asyncio.gather(...)
```